protocol/contracts/amo/TpfAmo.sol

```diff
@@ -0,0 +1,547 @@
+
+    event RecoveredToken(address, address, uint256);
+    event SetOperator(address);
+    event SetPostRebalanceSlippage(uint64);
+    event SetCooldown(uint64);
+    event SetRebalanceRateChange(uint256);
+    event SetPauseState(bool);
+    event StableDeposited(uint256, uint256);
+    event RebalanceUp(uint256, uint256);
+    event RebalanceDown(uint256, uint256);
+    event SetPoolHelper(address);
+    event SetMaxRebalanceAmount(uint256);
+    event WithdrawStable(uint256, uint256);
+
+    constructor(
+        address _balancerVault,
+        address _temple,
+        address _stable,
+        address _bptToken,
+        address _amoStaking,
+        address _booster,
+        uint64 _templeIndexInPool,
+        bytes32 _balancerPoolId
+    ) {
+        balancerVault = AMO__IBalancerVault(_balancerVault);
+        temple = IERC20(_temple);
+        stable = IERC20(_stable);
+        bptToken = IERC20(_bptToken);
+        amoStaking = AMO__IAuraStaking(_amoStaking);
+        booster = AMO__IAuraBooster(_booster);
+        templeBalancerPoolIndex = _templeIndexInPool;
+        balancerPoolId = _balancerPoolId;
+    }
+
+    function setPoolHelper(address _poolHelper) external onlyOwner {
+        poolHelper = AMO__IPoolHelper(_poolHelper);
+
+        emit SetPoolHelper(_poolHelper);
+    }
+
+    function setPostRebalanceSlippage(uint64 slippage) external onlyOwner {
+        if (slippage >= TPF_PRECISION || slippage == 0) {
+            revert AMOCommon.InvalidBPSValue(slippage);
+        }
+        postRebalanceSlippage = slippage;
+        emit SetPostRebalanceSlippage(slippage);
+    }
+
+    /**
+     * @notice Set maximum amount used by operator to rebalance
+     * @param maxAmount Maximum amount to set
+     */
+    function setMaxRebalanceAmount(uint256 maxAmount) external onlyOwner {
+        maxRebalanceAmount = maxAmount;
+        emit SetMaxRebalanceAmount(maxAmount);
+    }
+
+    /**
+     * @notice Set operator
+     * @param _operator New operator
+     */
+    function setOperator(address _operator) external onlyOwner {
+        operator = _operator;
+
+        emit SetOperator(_operator);
+    }
+
+    /**
+     * @notice Set cooldown time to throttle operator bot
+     * @param _seconds Time in seconds between operator calls
+     * */
+    function setCoolDown(uint64 _seconds) external onlyOwner {
+        cooldownSecs = _seconds;
+
+        emit SetCooldown(_seconds);
+    }
+    
+    /**
+     * @notice Toggle AMO pause
+     * */
+    function togglePause() external onlyOwner {
+        paused = !paused;
+
+        emit SetPauseState(paused);
+    }
+
+    /**
+     * @notice Recover any token from AMO
+     * @param token Token to recover
+     * @param to Recipient address
+     * @param amount Amount to recover
+     */
+    function recoverToken(address token, address to, uint256 amount) external onlyOwner {
+        if (to == address(0)) revert AMOCommon.InvalidAddress();
```

Comment:
This is checked in ERC20.sol. Given this is the only time this custom error is raised, probably not worth it here.